### PR TITLE
docs: add AGENTS.md and REVIEW.md

### LINKED ISSUE #11: Create AGENTS.md for AI coding agent context
## Summary

Create an `AGENTS.md` file at the repo root. This is a dedicated README for AI coding agents (Copilot, Devin, Codex, etc.) that provides predictable context and instructions.

Based on [agents.md](https://agents.md/) best practices.

### Project overview
- What silicon-bridge is: SNN-to-FPGA deployment pipeline
- Core crates: `fpga_export`, `fpga_metrics`, `fpga_bridge` (uart feature)
- License: dual MIT/Apache-2.0

### Dev environment tips
- `cargo check` — quick compile check
- `cargo test` — run all tests + doctests
- `cargo build --release` — optimized build
- Feature flags: `uart` for UART bridge support

### Testing instructions
- Tests live in `src/fpga_export.rs` (unit tests)
- Doctests in `src/lib.rs` Quick Start example
- Run `cargo test` before committing
- All tests must pass (3 unit + 1 doctest)

### Code style
- Rust edition 2024
- SPDX headers: `// SPDX-License-Identifier: MIT OR Apache-2.0`
- Doc comments with `//!` at module level
- `serde` derive for serializable types

### PR instructions
- Branch naming: `chore/`, `fix/`, `feat/` prefix
- Run `cargo test` and `cargo check` before pushing
- Keep PRs focused — one issue per PR when possible

### Security considerations
- No secrets, DSNs, or API keys in code
- UART feature requires `serialport` — optional dependency

## Acceptance Criteria

- [ ] `AGENTS.md` exists at repo root
- [ ] Covers project overview, dev tips, testing, code style, PR guidelines
- [ ] Accurate commands that work in this repo
- [ ] Keeps README.md clean for human contributors

### LINKED ISSUE #12: Create REVIEW.md with code review guidelines
## Summary

Create a `REVIEW.md` file at the repo root documenting code review guidelines for silicon-bridge. This ensures consistent review quality across contributors and AI reviewers.

### General principles
- Code must compile and pass all tests
- No secrets, credentials, or private telemetry in diffs
- Keep changes minimal and focused

#### Correctness
- [ ] Logic matches the stated intent
- [ ] Edge cases handled (empty vecs, zero values, overflow)
- [ ] Q8.8 conversions are correct and bounded

#### API design
- [ ] Public types use `pub` appropriately
- [ ] Builder pattern (e.g., `FpgaParameterExporter`) is ergonomic
- [ ] No unnecessary type restrictions (prefer `Real` over `Float32` for kwargs)

#### Documentation
- [ ] SPDX license header on all `.rs` files
- [ ] Doc comments on public items
- [ ] Examples in doc comments compile (doctests)

#### Testing
- [ ] New code has corresponding tests
- [ ] Tests cover happy path and edge cases
- [ ] Doctests are accurate

#### Security
- [ ] No hardcoded secrets or API keys
- [ ] No unsafe code without justification
- [ ] UART feature properly gated behind `#[cfg(feature = "uart")]`

### Common review comments
- Edition `2024` is valid — don't suggest downgrading to `2021`
- Relative license links break in rustdoc — use plain text
- Changelog entries need blank lines before lists (markdownlint)

### AI reviewer notes
- Codacy flags markdown lint issues — check CHANGELOG formatting
- Gemini may flag relative links — already addressed in lib.rs
- CodeRabbit auto-resolves when fix is pushed — verify thread status

### PR DESCRIPTION
**Pull request by rmems**
## Summary

- Add `AGENTS.md` — dedicated context for AI coding agents (project overview, dev tips, testing, code style, PR guidelines, security)
- Add `REVIEW.md` — code review checklist, common review comments, AI reviewer notes

Based on [agents.md](https://agents.md/) best practices.

Closes #11
Closes #12

---
Rovo Dev code review: <strong>Out of Rovo Dev credits</strong>
You've used all your Rovo Dev credits, so Rovo Dev can't review your pull requests.

___

## **CodeAnt-AI Description**
**Add AI agent and code review guidance for the repository**

### What Changed
- Added `AGENTS.md` with project context, setup commands, testing steps, code style, PR guidance, and security notes for AI coding agents
- Added `REVIEW.md` with a review checklist for correctness, API design, documentation, testing, and security
- Included common review reminders for known project-specific cases, such as edition 2024, license link formatting, changelog spacing, and unused dependencies

### Impact
`✅ Clearer AI-assisted contributions`
`✅ Fewer missed review checks`
`✅ Faster onboarding for new contributors`
<details><summary><strong>💡 Usage Guide</strong></summary>

### Checking Your Pull Request
Every time you make a pull request, our system automatically looks through it. We check for security issues, mistakes in how you're setting up your infrastructure, and common code problems. We do this to make sure your changes are solid and won't cause any trouble later.

### Talking to CodeAnt AI
Got a question or need a hand with something in your pull request? You can easily get in touch with CodeAnt AI right here. Just type the following in a comment on your pull request, and replace "Your question here" with whatever you want to ask:
<pre>
<code>@codeant-ai ask: Your question here</code>
</pre>
This lets you have a chat with CodeAnt AI about your pull request, making it easier to understand and improve your code.

#### Example
<pre>
<code>@codeant-ai ask: Can you suggest a safer alternative to storing this secret?</code>
</pre>

### Preserve Org Learnings with CodeAnt
You can record team preferences so CodeAnt AI applies them in future reviews. Reply directly to the specific CodeAnt AI suggestion (in the same thread) and replace "Your feedback here" with your input:
<pre>
<code>@codeant-ai: Your feedback here</code>
</pre>
This helps CodeAnt AI learn and adapt to your team's coding style and standards.

#### Example
<pre>
<code>@codeant-ai: Do not flag unused imports.</code>
</pre>

### Retrigger review
Ask CodeAnt AI to review the PR again, by typing:
<pre>
<code>@codeant-ai: review</code>
</pre>

### Check Your Repository Health
To analyze the health of your code repository, visit our dashboard at [https://app.codeant.ai](https://app.codeant.ai). This tool helps you identify potential issues and areas for improvement in your codebase, ensuring your repository maintains high standards of code health.

</details>

## Summary by CodeRabbit

* **Documentation**
  * Added an AI coding agent playbook with repo-specific conventions for Rust 2024, security/quality rules (e.g., no secrets, no `unsafe` without justification), feature-flag expectations, and recommended `cargo` workflows for build/test/lint.
  * Added code review guidance covering correctness (including numeric/edge-case considerations), API and documentation standards, test expectations (unit tests and doctests), and security/cleanup checklist items.

### CODE REVIEW

File: AGENTS.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+All 4 tests must pass. Add tests for any new code you write.
+
+## Code style
+
+- SPDX license header on every `.rs` file: `// SPDX-License-Identifier: MIT OR Apache-2.0`
+- Module-level doc comments with `//!`
+- Public items get `///` doc comments
+- Use `serde` derive for serializable types
+- Prefer `Real` type bounds over `Float32` for public kwargs (accepts f32, i32, etc.)
```
[the remark this answers is not in the document]
Comment (rmems) [reply AGENTS.md]: Fixed. `Real` type doesn't exist in this crate — updated to use `f32` consistently.

File: REVIEW.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+- [ ] Logic matches the stated intent
+- [ ] Edge cases handled (empty vecs, zero values, overflow)
+- [ ] Q8.8 conversions are correct and bounded (0.0–255.996 unsigned)
+- [ ] No off-by-one errors in array indexing
+
+### API design
+
+- [ ] Public types use `pub` appropriately — don't expose internals
+- [ ] Builder pattern (e.g., `FpgaParameterExporter`) is ergonomic
+- [ ] Prefer `Real` type bounds over `Float32` for public kwargs
```
[the remark this answers is not in the document]
Comment (rmems) [reply REVIEW.md]: Fixed. `Real` type doesn't exist in this crate — updated to use `f32` consistently.

File: REVIEW.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+Lists in CHANGELOG.md need a blank line before them:
+```markdown
+### Changed
+
+- Item here (not directly after heading)
+```
+
+### Unused dependencies
+
+Check if new dependencies are actually imported. `rand` is currently listed but unused — avoid adding more dead dependencies.
```
[the remark this answers is not in the document]
Comment (rmems) [reply REVIEW.md:73]: Correct — `rand` is a leftover dependency from the original Eagle-Lander extraction. Good catch by the docs.

File: REVIEW.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+- Keep changes minimal and focused — one concern per PR
+- Follow existing patterns in the codebase
+
+## Review checklist
+
+### Correctness
+
+- [ ] Logic matches the stated intent
+- [ ] Edge cases handled (empty vecs, zero values, overflow)
+- [ ] Q8.8 conversions are correct and bounded (0.0–255.996 unsigned)
```
[the remark this answers is not in the document]
Comment (rmems) [reply REVIEW.md]: Fixed. Added signed Q8.8 range (-128.0–127.996) to the checklist.

File: AGENTS.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (rmems) [reply AGENTS.md]: Fixed. `Real` doesn't exist — updated to reference `f32` which is what the codebase actually uses.

File: REVIEW.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (rmems) [reply REVIEW.md]: Fixed. `Real` doesn't exist — updated to reference `f32` which is what the codebase actually uses.

File: AGENTS.md [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+| `src/fpga_export.rs` | 3 unit tests | Q8.8 conversion, parameter export, memory calculation |
+| `src/lib.rs` | 1 doctest | Quick Start example |
+
+Run before every commit:
+
+```bash
+cargo test
+```
+
+All 4 tests must pass before pushing. Add tests for any new code you write.
```
[the remark this answers is not in the document]
Comment (rmems) [reply AGENTS.md]: Fixed. Replaced hard-coded count with generic "All tests must pass before pushing."

File: AGENTS.md [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+- `cargo build --features uart` — build with UART bridge (requires `serialport`)
+- `cargo clippy` — lint
+
+## Project overview
+
+`silicon-bridge` is a Rust crate for SNN (Spiking Neural Network)-to-FPGA deployment. It exports trained SNN parameters as Q8.8 fixed-point `.mem` files for Vivado synthesis, provides a UART bridge for runtime spike exchange, and parses Vivado timing reports for CI/CD gating.
+
+- **License**: dual MIT (Massachusetts Institute of Technology) / Apache-2.0
+- **Rust edition**: 2024
+- **Crate type**: library (`silicon_bridge`)
```
[the remark this answers is not in the document]
Comment (rmems) [reply AGENTS.md:58]: Fixed in commit d74c472. Added crate map with `fpga_export`, `fpga_metrics`, `fpga_bridge` and defined FPGA acronym.

File: AGENTS.md [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+## Constraints (mandatory)
+
+- Do not commit secrets, API keys, DSNs, or credentials
+- Do not add `unsafe` code without explicit safety justification
+- Do not downgrade Rust edition from 2024 to 2021
+- Do not add unused dependencies — check if a crate is actually imported before adding it
+- Do not use relative links to license files in doc comments (they break on docs.rs)
+
+## Security (mandatory)
+
```
[the remark this answers is not in the document]
Comment (rmems) [reply AGENTS.md]: Fixed in commit 8d3047d. Merged Security section into Constraints — removed duplicate, UART rule moved to Constraints.

File: REVIEW.md [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+- Keep changes minimal and focused — one concern per PR
+- Follow existing patterns in the codebase
+
+## Review checklist
+
+### Correctness
+
+- [ ] Logic matches the stated intent
+- [ ] Edge cases handled (empty vecs, zero values, overflow)
+- [ ] Q8.8 conversions are correct and bounded (unsigned: 0.0–255.996; signed: -128.0–127.996)
```
[the remark this answers is not in the document]
Comment (rmems) [reply REVIEW.md]: Fixed in commit bf5e5f2. Updated the checklist to match the actual code clamping in `fpga_bridge.rs:59` (-127.99..127.99).
